File: backend/src/Api/Common/Validate.cs
using System.Net.Mail;

namespace Api.Common;

/// ตรวจ input ที่ขอบระบบ · คืน null = ผ่าน, คืน string = ข้อความไทยที่เอาไปโชว์ได้เลย
/// แยกจาก Problems เพราะเป็นฟังก์ชันบริสุทธิ์ เทสต์ได้โดยไม่ต้องมี HTTP
public static class Validate
{
    public static string? Name(string? value, string what) => value?.Trim() switch
    {
        null or "" => $"กรุณากรอก{what}",
        { Length: > Limits.NameLength } => $"{what}ยาวเกิน {Limits.NameLength} ตัวอักษร",
        _ => null,
    };

    public static string? Email(string? value)
    {
        var email = value?.Trim();
        if (string.IsNullOrEmpty(email)) return "กรุณากรอกอีเมล";
        if (email.Length > Limits.NameLength) return $"อีเมลยาวเกิน {Limits.NameLength} ตัวอักษร";
        return MailAddress.TryCreate(email, out _) ? null : "รูปแบบอีเมลไม่ถูกต้อง";
    }

    /// รหัสนักเรียนของโรงเรียน · เป็นตัวชี้ว่าเป็นใคร ไม่ใช่ความลับ ตัวยืนยันตัวตนจริงคือบัญชี Google
    public static string? StudentCode(string? value)
    {
        var code = value?.Trim();
        if (string.IsNullOrEmpty(code)) return "กรุณากรอกรหัสนักเรียน";
        if (code.Length > Limits.StudentCodeLength) return $"รหัสนักเรียนยาวเกิน {Limits.StudentCodeLength} ตัวอักษร";
        return code.All(char.IsLetterOrDigit) ? null : "รหัสนักเรียนใช้ได้เฉพาะตัวเลขและตัวอักษร";
    }

    /// เลขที่ในห้อง
    public static string? No(int value) =>
        value is < 1 or > Limits.MaxStudentNo ? $"เลขที่ต้องอยู่ระหว่าง 1 ถึง {Limits.MaxStudentNo}" : null;

    public static string? Message(string? value) => value?.Trim() switch
    {
        null or "" => "กรุณากรอกข้อความ",
        { Length: > Limits.MessageLength } => $"ข้อความยาวเกิน {Limits.MessageLength} ตัวอักษร",
        _ => null,
    };

    /// คะแนนเต็มของรายการประเมิน
    public static string? MaxScore(decimal value) => value switch
    {
        <= 0 => "คะแนนเต็มต้องมากกว่า 0",
        > Limits.ScoreCeiling => $"คะแนนเต็มต้องไม่เกิน {Limits.ScoreCeiling:0.##}",
        // DB เก็บ decimal(6,2) ถ้าไม่ดักตรงนี้ 10.555 จะถูกปัดเงียบ ๆ เป็น 10.56
        _ when value != Math.Round(value, 2) => "คะแนนเต็มมีทศนิยมได้ไม่เกิน 2 ตำแหน่ง",
        _ => null,
    };

    /// คะแนนที่กรอก · null = ยังไม่ให้คะแนน ถือว่าผ่าน
    public static string? Score(decimal? value, decimal maxScore) => value switch
    {
        null => null,
        < 0 => "คะแนนติดลบไม่ได้",
        _ when value > maxScore => $"คะแนนเกินคะแนนเต็ม ({maxScore:0.##})",
        // DB เก็บ decimal(6,2) ถ้าไม่ดักตรงนี้ 18.555 จะถูกปัดเงียบ ๆ เป็น 18.56 (ใช้ทั้งกรอกเองและนำเข้า Excel)
        _ when value != Math.Round(value.Value, 2) => "คะแนนมีทศนิยมได้ไม่เกิน 2 ตำแหน่ง",
        _ => null,
    };
}

File: frontend/src/lib/validate.ts
/**
 * ตรวจ input ฝั่งเว็บ — คืน null = ผ่าน, คืน string = ข้อความไทยที่โชว์ได้เลย
 *
 * ตั้งใจให้ "หน้าตาเหมือน" Common/Validate.cs ฝั่ง backend เป๊ะ ๆ
 * ฝั่งนี้มีไว้ให้ผู้ใช้รู้ผลทันทีโดยไม่ต้องรอ server — **ไม่ใช่** ตัวกันจริง
 * ตัวกันจริงคือฝั่ง backend เสมอ ห้ามตัดออกเด็ดขาด
 */

/** ต้องตรงกับ Common/Limits.cs */
export const LIMITS = {
  name: 200,
  message: 2000,
  studentCode: 20,
  scoreCeiling: 9999.99,
} as const

const trimmed = (value: string | null | undefined) => (value ?? '').trim()

/** DB เก็บ decimal(6,2) · ปัดด้วย 100 แล้วเทียบ กันเศษทศนิยมของ float แบบ 0.1 + 0.2 */
const hasAtMost2Decimals = (value: number) => Math.abs(Math.round(value * 100) - value * 100) < 1e-6

export const validate = {
  name(value: string | null | undefined, what: string): string | null {
    const v = trimmed(value)
    if (!v) return `กรุณากรอก${what}`
    if (v.length > LIMITS.name) return `${what}ยาวเกิน ${LIMITS.name} ตัวอักษร`
    return null
  },

  email(value: string | null | undefined): string | null {
    const v = trimmed(value)
    if (!v) return 'กรุณากรอกอีเมล'
    if (v.length > LIMITS.name) return `อีเมลยาวเกิน ${LIMITS.name} ตัวอักษร`
    // เช็คหยาบ ๆ พอให้รู้ตัวเร็ว ตัวตัดสินจริงคือ MailAddress ฝั่ง backend
    if (!/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(v)) return 'รูปแบบอีเมลไม่ถูกต้อง'
    return null
  },

  studentCode(value: string | null | undefined): string | null {
    const v = trimmed(value)
    if (!v) return 'กรุณากรอกรหัสนักเรียน'
    if (v.length > LIMITS.studentCode) return `รหัสนักเรียนยาวเกิน ${LIMITS.studentCode} ตัวอักษร`
    if (!/^[A-Za-z0-9]+$/.test(v)) return 'รหัสนักเรียนใช้ได้เฉพาะตัวเลขและตัวอักษร'
    return null
  },

  message(value: string | null | undefined): string | null {
    const v = trimmed(value)
    if (!v) return 'กรุณากรอกข้อความ'
    if (v.length > LIMITS.message) return `ข้อความยาวเกิน ${LIMITS.message} ตัวอักษร`
    return null
  },

  /** คะแนนเต็มของรายการประเมิน */
  maxScore(value: number | null | undefined): string | null {
    if (value == null || Number.isNaN(value)) return 'กรุณากรอกคะแนนเต็ม'
    if (value <= 0) return 'คะแนนเต็มต้องมากกว่า 0'
    if (value > LIMITS.scoreCeiling) return `คะแนนเต็มต้องไม่เกิน ${LIMITS.scoreCeiling}`
    if (!hasAtMost2Decimals(value)) return 'คะแนนเต็มมีทศนิยมได้ไม่เกิน 2 ตำแหน่ง'
    return null
  },

  /** คะแนนที่กรอก · null = ยังไม่ให้คะแนน ถือว่าผ่าน */
  score(value: number | null | undefined, maxScore: number): string | null {
    if (value == null) return null
    if (Number.isNaN(value)) return 'คะแนนต้องเป็นตัวเลข'
    if (value < 0) return 'คะแนนติดลบไม่ได้'
    if (value > maxScore) return `คะแนนเกินคะแนนเต็ม (${maxScore})`
    if (!hasAtMost2Decimals(value)) return 'คะแนนมีทศนิยมได้ไม่เกิน 2 ตำแหน่ง'
    return null
  },
}

/** คืนข้อความแรกที่ไม่ผ่าน หรือ null ถ้าผ่านหมด */
export const firstError = (...checks: (string | null)[]): string | null => checks.find(Boolean) ?? null
